docs/source/notes/gradcheck.rst

```diff
@@ -0,0 +1,221 @@
+This section presents a way to perform gradcheck in a faster way without affecting its correctness.
+The debuggability can be recovered by adding special code logic when we detect an error. In that case, we can run the slow code that reconstruct the full matrix to give full details to the user.
+
+The high level strategy here is to find a scalar quantity that can be computed efficiently by both the numerical and analytical methods and that represents the full matrix computed by the slow gradcheck well enough to ensure that it will catch any discrepancy in the Jacobians.
+
+Fast gradcheck for functions in :math:`\mathcal{R}^N \to \mathcal{R}^M`
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The scalar quantity that we want to compute here is :math:`v^T J_f u` for a given random vector :math:`v \in \mathcal{R}^M` and a random unit norm vector :math:`u \in \mathcal{R}^N`.
+
+For the numerical evaluation, we can efficiently compute :math:`J_f u \approx \frac{f(x + u * eps) - f(x - u * eps)}{2 * eps}`. We then perform the dot product between this vector and :math:`v` to get the scalar value of interest.
+
+For the analytical version, we can use backward mode AD to compute :math:`v^T J_f` directly. We then perform the dot product with :math:`u` to get the expected value.
+
+Fast gradcheck for functions in :math:`\mathcal{C}^N \to \mathcal{R}^M`
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In a similar way to the real-to-real case, we want to perform a reduction of the full matrix. But the :math:`2 * CW` matrix is complex-valued and so in this case, we will compare to complex scalars.
+
+Due to some constraints on what we can compute efficiently in the numerical case and to keep the number of numerical evaluations to a minimum, we compute the following (albeit surprising) scalar value:
+
+.. math::
+    s := 2 * v^T (real(CW) ur + i * imag(CW) ui)
+
+where :math:`v \in \mathcal{R}^M`, :math:`ur \in \mathcal{R}^N` and :math:`ui \in \mathcal{R}^N`.
+
+Fast complex input numerical evaluation
+"""""""""""""""""""""""""""""""""""""""
+
+We first consider how to compute :math:`s` with a numerical method. To do so, keeping in mind that we're considering :math:`g: \mathcal{C}^N \to \mathcal{R}^M, z \to y` with :math:`z = a + i b`, and that :math:`CW = \frac{1}{2} * (\frac{\partial y}{\partial a} + i \frac{\partial y}{\partial b})`,  we rewrite it as follows:
+
+.. math::
+    \begin{aligned}
+        s &= 2 * v^T (real(CW) ur + i * imag(CW) ui)
+          &= 2 * v^T (\frac{1}{2} * \frac{\partial y}{\partial a} ur + i * \frac{1}{2} * \frac{\partial y}{\partial b} ui)
+          &= v^T (\frac{\partial y}{\partial a} ur + i * \frac{\partial y}{\partial b} ui)
+          &= v^T ((\frac{\partial y}{\partial a} ur) + i * (\frac{\partial y}{\partial b} ui))
+    \end{aligned}
+
```

Comment:
![image](https://user-images.githubusercontent.com/5652049/115565611-83c5e000-a287-11eb-8963-2ab8ee3694db.png)

nit: This gets cut off (you can scroll to the right to recover it, though). It might look nicer if everything were on different lines?

Reply:
This is definitely missing the end of lines! Good catch!